source/ceylon/logging/LoggerFactory.ceylon

```diff
@@ -0,0 +1,90 @@
+import ceylon.language.meta.model {
+    ClassOrInterface
+}
+import ceylon.collection {
+    MutableMap,
+    HashMap
+}
+
+"`LoggerFactory` can be satisfied to define a custom
+ logging strategy. Typically, an instance of `LoggerFactory`
+ will be be assigned to [[loggerFactory]]."
+shared
+interface LoggerFactory {
+
+    "Function to obtain a [[Logger]] for the given
+     [[Category]].
+
+     Callers intending to wrap the returned logger should
+     provide the [[ClassOrInterface]] of the wrapper.
+     For example:
+
+            FormattingLogger formattingLogger(Category category)
+                =>  FormattingLogger(loggerFactory.logger(
+                        category, `FormattingLogger`));"
+    shared formal
+    Logger logger<Wrapper>(
+            Category category,
+            "The class that will be used to wrap the returned
+             value, if any. This may be used by loggers that
+             log source file and line number information."
+            ClassOrInterface<Wrapper>? wrapper=null)
+            given Wrapper satisfies Object;
+}
```

Comment:
I believe it was to have a generic method, so I'd have a T to use for `javaClass<T>().name`.

A secondary reason would be for future support to abstract nested or mapped diagnostic contexts, at which time there'd be more than one method. I've been thinking about this for a while, but not long enough to make a request or recommendation. But I do know that they can be very useful server side.

There still is of course the simplified option I mentioned in the PR. It turns out that `javaClass<T>().name` is fast enough to call for every (enabled) call to the wrapped logger, at least per my testing.